### DISCUSSION TITLE
Filesystem api

### DISCUSSION THREAD

**Discussion by parasharsangeet** [orden-devel/orden#5] [Ideas]
I think we should start with the filesystem api to open files and folders.

**Comment by shahab96** [comment]
We should probably get the backend for that available first because the front end will actually need a few pieces there. The file tree needs to be able to be read which I assume we want the backend to inject into the front end (?)

We may need to work on the back and front end in parallel as two separate issues.

Additionally, what kind of api do we want to expose? Like are we just sending file metadata or is this going to include file contents as well? Personally I think the file metadata and file content apis should be separate.

**Reply by parasharsangeet** [reply]
Tauri has things called commands which can be defined in rust. With this we can invoke the rust command when the user clicks on the open button. We can access the filesystem from the backend(rust) which can send the required details of the files. And I think segregating the apis will be helpful because we would be able to lazy load the contents of the file while still loading the metadata.

**Reply by shahab96** [reply]
I agree with segregating the apis. I do wonder how we will implement the communication for the filesystem api though. Like I imagine something along the lines of, backend scans the directory and sends the file tree data to the front end which generates a bunch of buttons. User clicks on a button and that initiates a request to the backend which then opens said file. Perhaps something like that?

**Reply by parasharsangeet** [reply]
Yep! clicking on the button will initiate an api request which we can process and send the data to the frontend.

**Reply by shahab96** [reply]
Does this mean we need to create a startup process first, in order to boot the app and perform the scan on the filesystem so that the front end actually knows what to display?

**Reply by parasharsangeet** [reply]
Why would we scan the filesystem on boot?

**Reply by shahab96** [reply]
Not on machine boot. On application startup

The frontend needs to display something to the user, I just assume it would be the file tree for the directory they decided to open the text editor for.

**Reply by parasharsangeet** [reply]
We could keep a cache file where we can store the path of the folder the last time the user opened the editor. If its none we can show them a startup page with like Open Folder, Open file, etc.

**Reply by shahab96** [reply]
Fair enough.

**Reply by Avondsch** [reply]
am good with sangeet's approach

**Comment by shahab96** [comment]
Let's get an API definition set up here before we proceed with creating issues. @acidicity @Kofee69

**Reply by Avondsch** [reply]
cool

**Comment by parasharsangeet** [comment]
We have to make a decision here. In tauri we have two ways of communicating to the backend. One is through the traditional command system and the other is through events. Through commands there is just a one way communication. You can think of it like http requests. However, events allow bidirectional communication. Essentially we can emit events from the backend as well as the frontend. We also can listen for events.

**Reply by parasharsangeet** [reply]
For the filesystem api, I don't think so we need to use events. Since, we are just going to make an api call and we get some data. This doesn't require events from what i think.
